rust/kernel/net/skbuff.rs

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Definitors for the struct [`crate::bindings::sk_buff`] memory handlers.
+
+use core::{ops::Drop, ptr};
+
+use crate::bindings;
+use crate::types::SavedAsPointer;
+
+/// Wraps the kernel's `struct sk_buff`.
+///
+/// # Invariants
+///
+/// The pointer `Self::ptr` is non-null and valid.
+#[repr(transparent)]
+pub struct SkBuff {
+    ptr: *const bindings::sk_buff,
+}
+
+impl SkBuff {
+    /// Binding to `bindings::skb_clone_tx_timestamp`.
+    ///
+    /// Noop if `CONFIG_NETWORK_PHY_TIMESTAMPING` is not enabled.
+    #[cfg(CONFIG_NETWORK_PHY_TIMESTAMPING)]
+    pub fn clone_tx_timestamp(&mut self) {
+        // SAFETY: self.ptr is valid if self is valid.
+        unsafe {
+            bindings::skb_clone_tx_timestamp(self.ptr as *mut bindings::sk_buff);
+        }
+    }
+
+    /// Binding to `bindings::skb_clone_tx_timestamp`.
+    ///
+    /// Noop if `CONFIG_NETWORK_PHY_TIMESTAMPING` is not enabled.
+    #[cfg(not(CONFIG_NETWORK_PHY_TIMESTAMPING))]
+    pub fn clone_tx_timestamp(&mut self) {
+        // NOOP
+    }
+
+    /// Driver hook for transmit timestamping.
+    ///
+    /// Ethernet MAC Drivers should call this function in their hard_xmit()
+    /// function immediately before giving the sk_buff to the MAC hardware.
+    ///
+    /// Specifically, one should make absolutely sure that this function is
+    /// called before TX completion of this packet can trigger.  Otherwise
+    /// the packet could potentially already be freed.
+    pub fn tx_timestamp(&mut self) {
+        self.clone_tx_timestamp();
+        if self.shinfo().tx_flags() as u32 & bindings::SKBTX_SW_TSTAMP != 0 {
+            unsafe {
+                bindings::skb_tstamp_tx(self.ptr as *mut bindings::sk_buff, ptr::null_mut());
+            }
+        }
+    }
+
+    /// Length of the [`SkBuff`].
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> u32 {
+        self.get_internal().len
+    }
+
+    /// Get the Shared info for this SKBuffer.
```

Comment:
I would avoid repeating "for this SKBuffer", "for Self", etc. if they are methods.

Also, not sure why `Shared` is capitalized.

```suggestion
    /// Get shared information.
```